### PR TITLE
Merge with PostgreSQL 9.3

### PR DESCRIPTION
**Pull request by hlinnaka**
I'm opening this PR as a heads up that we're about to complete the 9.3 merge, and land this on the GPDB master branch. This will be squashed into a single Merge commit before pushing. Please take a look and comment on the proposed commit message below:

```

Merge with PostgreSQL 9.3 (up to almost 9.3beta2)

Merge with PostgreSQL, up to the point where the REL9_3_STABLE branch
was created, and 9.4 development started on the PostgreSQL master
branch. That is almost up to 9.3beta2.

Notable upstream changes, from a GPDB point of view:

* LATERAL support. Mostly works in GPDB now, although performance
  might not be very good. LATERAL subqueries, except for degenerate
  cases that can be made non-LATERAL during optimization, typically
  use nested loop joins. Unless the data distribution is the same on
  both sides of the join, GPDB needs to add Motion nodes, and cannot
  push down the outer query parameter to the inner side through the
  motion. That is the same problem we have with SubPlans and nested
  loop joins in general, but it happens frequently with LATERAL. Also,
  there are a couple of cases, covered by the upstream regression
  tests, where the planner currently throws an error. They have been
  disabled and marked with GPDB_93_MERGE_FIXME comments, and will need
  to be investigated later. Also, no ORCA support for LATERAL yet.

* Materialized views. They have not been made to work in GPDB
  yet. CREATE MATERIALIZED VIEW works, but REFRESH MATERIALIZED VIEW
  does not. The 'matviews' test has been temporarily disabled, until
  that's fixed. There is a GPDB_93_MERGE_FIXME comment about this too.

* Support for background worker processes. Nothing special was done
  about them in the merge, but we could now make use of them for all
  the various GPDB-specific background processes, like the FTS prober
  and gpmon processes.

* Support for writable foreign tables was introduced. I believe
  foreign tables now have all the same functionality, at a high level,
  as external tables, so we could start merging the two concepts. But
  this merge commit doesn't do anything about that yet, external
  tables and foreign tables are still two entirely different beasts.

* A lot of expected output churn, thanks to a few upstream changes. We
  no longer print a NOTICE on implicitly created indexes and sequences
  (commit d7c734841b), and the rules on when table aliases are printed
  were changed (commit 11e131854f).

* Caught up to a bunch of features that we had already backported from
  9.3: data page checksums, numeric datatype speedups, COPY FROM/TO
  PROGRAM, and pg_upgrade as whole.


A couple of other noteworthy changes:

* contrib/xlogdump utility is removed, in favor of the upstream
  contrib/pg_xlogdump utility.

* Removed "idle session timeout" hook. The current implementation was
  badly broken by upstream refactoring of timeout handling (commit
  f34c68f096). We'll probably need to re-introduce it in some form,
  but it will probably look quite different, to make it fit more
  nicely with the new timeout APIs.

Co-authored-by: Ashwin Agrawal <aagrawal@pivotal.io>
Co-authored-by: Asim R P <apraveen@pivotal.io>
Co-authored-by: David Kimura <dkimura@pivotal.io>
Co-authored-by: Ekta Khanna <ekhanna@pivotal.io>
Co-authored-by: Heikki Linnakangas <hlinnakangas@pivotal.io>
Co-authored-by: Jacob Champion <pchampion@pivotal.io>
Co-authored-by: Jinbao Chen <jinchen@pivotal.io>
Co-authored-by: Kalen Krempely <kkrempely@pivotal.io>
Co-authored-by: Paul Guo <paulguo@gmail.com>
Co-authored-by: Richard Guo <guofenglinux@gmail.com>
Co-authored-by: Shaoqi Bai <sbai@pivotal.io>
```

### CODE REVIEW

**Comment by ivannovick** [comment]
WOW!  Amazing work

File: contrib/pg_upgrade/pg_upgrade.h [hunk 1]
```diff
@@ -125,10 +125,8 @@ extern char *output_files[];
  * ("Improve concurrency of foreign key locking") which also updated catalog
  * version to this value.  pg_upgrade behavior depends on whether old and new
  * server versions are both newer than this, or only the new one is.
- *
- * GPDB_93_MERGE_FIXME
-#define MULTIXACT_FORMATCHANGE_CAT_VER 201301231
  */
+#define MULTIXACT_FORMATCHANGE_CAT_VER 201301231
```
Comment (jchampio) [inline contrib/pg_upgrade/pg_upgrade.h]: This catalog version should be GPDB-ified, I think. (I assume that should apply to the other version numbers here as well, but those don't need to happen in this merge.)

**Comment by hlinnaka** [comment]
On 20/09/18 18:37, Jacob Champion wrote:
> *@jchampio* commented on this pull request.
> 
> ------------------------------------------------------------------------
> 
> In contrib/pg_upgrade/pg_upgrade.h 
> <https://github.com/greenplum-db/gpdb/pull/5805#discussion_r219212773>:
> 
>>   */
> +#define MULTIXACT_FORMATCHANGE_CAT_VER 201301231
> 
> This catalog version should be GPDB-ified, I think. (I assume that 
> should apply to the other version numbers here as well, but those don't 
> need to happen in this merge.)

Hmm. We haven't actually changed the format, it's exactly the same as in 
upstream, I think. I don't think we have any coherent policy on when to 
change these.

- Heikki

**Comment by jchampio** [comment]
> Hmm. We haven't actually changed the format, it's exactly the same as in
upstream, I think. I don't think we have any coherent policy on when to
change these.

@hlinnaka I think GPDB 4 and 5 already have catalog versions later than `201301231`, which means pg_upgrade will assume they have `MultiXact` data and bail out when they don't. The policy probably needs to be "always change the catalog versions to the GPDB equivalents", just like we change the version boundaries in pg_dump.

**Comment by hlinnaka** [comment]
On 20/09/18 19:10, Jacob Champion wrote:
>     Hmm. We haven't actually changed the format, it's exactly the same as in
>     upstream, I think. I don't think we have any coherent policy on when to
>     change these.
> 
> @hlinnaka <https://github.com/hlinnaka> I think GPDB 4 and 5 already 
> have catalog versions later than |201301231|, which means pg_upgrade 
> will assume they have |MultiXact| data and bail out when they don't. The 
> policy probably needs to be "always change the catalog versions to the 
> GPDB equivalents", just like we change the version boundaries in pg_dump.

Oh, I see. I didn't understand what that constant is used for, until 
now. Yes, we should change it.

- Heikki

**Comment by hlinnaka** [comment]
Pushed! Thanks to everyone involved!

**Pull request closed without merging**